daffodil-test/src/test/scala/org/apache/daffodil/infoset/TestStringAsXml.scala

```diff
@@ -0,0 +1,193 @@
+    val parseStrOpt = if (parseRes.isProcessingError) None else Some(parseBos.toString)
+    (parseDiags, parseStrOpt)
+  }
+
+  private def doUnparse(dp: DataProcessor, infoset: InputStream) = {
+    val unparseIn = new XMLTextInfosetInputter(infoset)
+    val unparseBos = new ByteArrayOutputStream()
+    val unparseOut = java.nio.channels.Channels.newChannel(unparseBos)
+    val unparseRes = dp.unparse(unparseIn, unparseOut)
+    val unparseDiags = unparseRes.getDiagnostics.map(_.toString)
+    val unparseStrOpt = if (unparseRes.isProcessingError) None else Some(unparseBos.toString)
+    (unparseDiags, unparseStrOpt)
+  }
+
+  @Test def test_stringAsXml_01(): Unit = {
+    val dp = compileSchema(Misc.getRequiredResource("/org/apache/daffodil/infoset/stringAsXml/namespaced/xsd/binMessage.dfdl.xsd"))
+    val parseData = Misc.getRequiredResource("/org/apache/daffodil/infoset/stringAsXml/namespaced/binMessage_01.dat").toURL.openStream
+    val (parseDiags, Some(parseInfosetActual)) = doParse(dp, parseData)
+    val parseInfosetExpected = {
+      val is = Misc.getRequiredResource("/org/apache/daffodil/infoset/stringAsXml/namespaced/binMessage_01.dat.xml").toURL.openStream
+      IOUtils.toString(is, StandardCharsets.UTF_8)
+    }
+    // diagnostic from full validation
+    assertTrue(parseDiags.find(_.contains("Element 'xmlStr' is a simple type")).isDefined)
+    // diagnostic from limited validation
+    assertTrue(parseDiags.find(_.contains("xmlStr failed facet checks due to: facet maxLength")).isDefined)
+    // we still get the expected infoset, replace CRLF with LF because of git windows autocrlf
+    assertEquals(parseInfosetExpected.replace("\r\n", "\n"), parseInfosetActual.replace("\r\n", "\n"))
+
+    // validate the infoset using the handwritten WithPayload schema
+    val xsdFile = new File(Misc.getRequiredResource("/org/apache/daffodil/infoset/stringAsXml/namespaced/xsd/binMessageWithXmlPayload.xsd"))
+    val xsdFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI)
+    val xsd = xsdFactory.newSchema(xsdFile)
+    val source = new StreamSource(new ByteArrayInputStream(parseInfosetActual.getBytes(StandardCharsets.UTF_8)))
+    val validator = xsd.newValidator()
+    val e = intercept[SAXParseException] {
+      validator.validate(source)
+    }
+    assertTrue(e.toString.contains("Value '=invalid field' is not facet-valid"))
+  }
+
+  @Test def test_stringAsXml_02(): Unit = {
+    val dp = compileSchema(Misc.getRequiredResource("/org/apache/daffodil/infoset/stringAsXml/namespaced/xsd/binMessage.dfdl.xsd"))
+    val unparseInfoset = Misc.getRequiredResource("/org/apache/daffodil/infoset/stringAsXml/namespaced/binMessage_01.dat.xml").toURL.openStream
+    val (_, Some(unparseDataActual)) = doUnparse(dp, unparseInfoset)
+    val unparseDataExpected = {
+      val is = Misc.getRequiredResource("/org/apache/daffodil/infoset/stringAsXml/namespaced/binMessage_01.dat.xml.dat").toURL.openStream
```

Comment:
Why the funny file name .dat.xml.dat? Are you nesting these things? If so deserves a comment. If not, ordinary naming would be better.

Reply:
It's my way of saying the original is `foo.dat`, the expected infoset created when parsing that file is `foo.dat.xml`, and the expected data when unparsing that is `foo.dat.xml.dat`. So `foo.dat` and `foo.dat.xml.dat` are functionally the same, but the latter has canonicalized XML in the payload.